Detectors/TRD/calibration/src/CalibratorGain.cxx

```diff
@@ -0,0 +1,202 @@
+
+using namespace o2::trd::constants;
+
+namespace o2::trd
+{
+
+using Slot = o2::calibration::TimeSlot<GainCalibHistos>;
+
+void CalibratorGain::initOutput()
+{
+  // reset the CCDB output vectors
+  mInfoVector.clear();
+  mObjectVector.clear();
+}
+
+void CalibratorGain::initProcessing()
+{
+  if (mInitDone) {
+    return;
+  }
+  for (int iDet = 0; iDet < MAXCHAMBER; ++iDet) {
+    mdEdxhists[iDet] = std::make_unique<TH1F>(Form("hdEdx%d", iDet), "dEdx", NBINSGAINCALIB, 0., NBINSGAINCALIB);
+  }
+
+  // set tree addresses
+  if (mEnableOutput && mOutTree) {
+    mOutTree->Branch("MPVdEdx", &fitResults);
+    for (int iDet = 0; iDet < MAXCHAMBER; ++iDet) {
+      mOutTree->Branch(Form("dEdxHist_%d", iDet), mdEdxhists[iDet].get());
+    }
+  }
+
+  mInitDone = true;
+}
+
+void CalibratorGain::retrievePrev(o2::framework::ProcessingContext& pc)
+{
+  static bool doneOnce = false;
+  if (!doneOnce) {
+    doneOnce = true;
+    fitResults.fill(constants::MPVDEDXDEFAULT);
+    // We either get a pointer to a valid object from the last ~hour or to the default object
+    // which is always present. The first has precedence over the latter.
+    /*auto dataCalGain = pc.inputs().get<o2::trd::CalGain*>("calgain");
+    std::string msg = "Default Object";
+    // We check if the object we got is the default one by comparing it to the defaults.
+    for (int iDet = 0; iDet < MAXCHAMBER; ++iDet) {
+      if (dataCalGain->getMPVdEdx(iDet) != constants::MPVDEDXDEFAULT) {
+        msg = "Previous Object";
+        break;
+      }
+    }
+    LOG(info) << "Calibrator: From CCDB retrieved " << msg;
+
+    // Here we set each entry regardless if it is the default or not.
+    for (int iDet = 0; iDet < MAXCHAMBER; ++iDet) {
+      fitResults[iDet] = dataCalGain->getMPVdEdx(iDet);
+    }*/
+  }
+}
+
+void CalibratorGain::finalizeSlot(Slot& slot)
+{
+  // do actual calibration for the data provided in the given slot
+  TStopwatch timer;
+  timer.Start();
+  initProcessing();
+
+  auto dEdxHists = slot.getContainer();
+  for (int iDet = 0; iDet < MAXCHAMBER; ++iDet) {
+    for (int iBin = 0; iBin < NBINSGAINCALIB; ++iBin) {
+      mdEdxhists[iDet]->SetBinContent(iBin + 1, dEdxHists->getHistogramEntry(iDet * NBINSGAINCALIB + iBin));
+    }
+    int nEntries = mdEdxhists[iDet]->Integral();
+    // Check if we have the minimum amount of entries
+    if (nEntries < mMinEntriesChamber) {
+      LOGF(debug, "Chamber %d did not reach minimum amount of %d entries for refit", iDet, mMinEntriesChamber);
+      continue;
+    }
+    mdEdxhists[iDet]->Scale(1. / nEntries);
+
+    // Fitting histogram
+    TF1Convolution* fConv = new TF1Convolution("[3] * TMath::Landau(x, [0], [1]) * exp(-[2]*x)", "TMath::Gaus(x, 0, [0])", 0., NBINSGAINCALIB, true);
+    fConv->SetNofPointsFFT(2000);
+    TF1* fitFunction = new TF1("fitConvLandau", *fConv, 0., NBINSGAINCALIB, fConv->GetNpar());
+    fitFunction->SetParameters(mdEdxhists[iDet]->GetMean() / 1.25, 15, 0.02, 1, 0.1);
```

Comment:
Btw, how stable are your fits? I struggled somewhat with this and achieved stability with setting parameter limits like these here:
``` C
     fitFunction->SetParLimits(0, 30., 100.0);
      fitFunction->SetParLimits(1, 5.0, 25.0);
      fitFunction->SetParLimits(2, -0.1, 0.5);
      fitFunction->SetParLimits(3, -0.1, 0.5);
      fitFunction->SetParLimits(5, 1.0, 10.0);
```